elastic/apm-server
# Add fingerprint calculation for servicemaps to spans.

## Motivation/summary
Calculate span fingerprint based on `service`, `destination`, `type` and `subtype` information and index as `span.servicemap.fingerprint`.

Adding a dedicated field calculated from the above mentioned fields allows for more performant processing of the data for the new servicemaps feature in Kibana. See detailed motivation description in #3335 .

## Checklist

~- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).~

~- [ ] I have made corresponding changes to the documentation~


## How to test these changes
Ingest spans and ensure a `span.servicemap.fingerprint` is stored accordingly: 
* The fingerprint is the same for events containing the same information for 
`service.name`, `service.environment`, `span.type`, `span.subtype`, `span.destination`
* No fingerprint value is set for spans where no `span.destination.address` is given.
* for spans where `service` information is overwritten on a per event base, the overwritten service information is used for the calculation.

The testfile [intake-v2/spans.ndjson](https://github.com/elastic/apm-server/blob/master/testdata/intake-v2/spans.ndjson) contains data covering scenarios with and without destination information and with and without overriding metadata service information.

## Related issues
Closes #3335
